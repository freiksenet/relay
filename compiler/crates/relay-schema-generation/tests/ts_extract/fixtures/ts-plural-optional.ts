/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */


//extract
function plural_string(user: User): Array<string> {}

//extract
function plural_optional_string(user: User): Array<string | null | undefined> {}

//extract
function optional_plural_string(user: User): Array<string> | null | undefined {}

//extract
function optional_plural_optional_string(user: User): Array<string | null | undefined> | null | undefined {}

function ignored(user: User): Array<string | null | undefined> | null | undefined {}

// Multiple top comments
// this another
function ignored(user: User): Array<string | null | undefined> | null | undefined {}