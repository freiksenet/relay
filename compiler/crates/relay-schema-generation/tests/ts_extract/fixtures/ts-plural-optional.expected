==================================== INPUT ====================================
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */


//extract
function plural_string(user: User): Array<string> {}

//extract
function plural_optional_string(user: User): Array<string | null | undefined> {}

//extract
function optional_plural_string(user: User): Array<string> | null | undefined {}

//extract
function optional_plural_optional_string(user: User): Array<string | null | undefined> | null | undefined {}

function ignored(user: User): Array<string | null | undefined> | null | undefined {}

// Multiple top comments
// this another
function ignored(user: User): Array<string | null | undefined> | null | undefined {}
==================================== OUTPUT ===================================
Strong(
    FieldData {
        field_name: WithLocation {
            location: <generated>:212:225,
            item: "plural_string",
        },
        return_type: GenericTypeAnnotation(
            GenericTypeAnnotation {
                id: Identifier(
                    Identifier {
                        name: "Array",
                        binding: None,
                        type_annotation: None,
                        loc: None,
                        range: SourceRange {
                            start: 239,
                            end: 244,
                        },
                    },
                ),
                type_parameters: Some(
                    TypeParameterInstantiation {
                        params: [
                            StringTypeAnnotation(
                                StringTypeAnnotation {
                                    loc: None,
                                    range: SourceRange {
                                        start: 245,
                                        end: 251,
                                    },
                                },
                            ),
                        ],
                        loc: None,
                        range: SourceRange {
                            start: 244,
                            end: 252,
                        },
                    },
                ),
                loc: None,
                range: SourceRange {
                    start: 239,
                    end: 252,
                },
            },
        ),
        entity_type: Some(
            GenericTypeAnnotation(
                GenericTypeAnnotation {
                    id: Identifier(
                        Identifier {
                            name: "User",
                            binding: None,
                            type_annotation: None,
                            loc: None,
                            range: SourceRange {
                                start: 232,
                                end: 236,
                            },
                        },
                    ),
                    type_parameters: None,
                    loc: None,
                    range: SourceRange {
                        start: 232,
                        end: 236,
                    },
                },
            ),
        ),
        arguments: None,
        is_live: None,
    },
)

Strong(
    FieldData {
        field_name: WithLocation {
            location: <generated>:276:298,
            item: "plural_optional_string",
        },
        return_type: GenericTypeAnnotation(
            GenericTypeAnnotation {
                id: Identifier(
                    Identifier {
                        name: "Array",
                        binding: None,
                        type_annotation: None,
                        loc: None,
                        range: SourceRange {
                            start: 312,
                            end: 317,
                        },
                    },
                ),
                type_parameters: Some(
                    TypeParameterInstantiation {
                        params: [
                            UnionTypeAnnotation(
                                UnionTypeAnnotation {
                                    types: [
                                        StringTypeAnnotation(
                                            StringTypeAnnotation {
                                                loc: None,
                                                range: SourceRange {
                                                    start: 318,
                                                    end: 324,
                                                },
                                            },
                                        ),
                                        NullLiteralTypeAnnotation(
                                            NullLiteralTypeAnnotation {
                                                loc: None,
                                                range: SourceRange {
                                                    start: 327,
                                                    end: 331,
                                                },
                                            },
                                        ),
                                        GenericTypeAnnotation(
                                            GenericTypeAnnotation {
                                                id: Identifier(
                                                    Identifier {
                                                        name: "undefined",
                                                        binding: None,
                                                        type_annotation: None,
                                                        loc: None,
                                                        range: SourceRange {
                                                            start: 334,
                                                            end: 343,
                                                        },
                                                    },
                                                ),
                                                type_parameters: None,
                                                loc: None,
                                                range: SourceRange {
                                                    start: 334,
                                                    end: 343,
                                                },
                                            },
                                        ),
                                    ],
                                    loc: None,
                                    range: SourceRange {
                                        start: 318,
                                        end: 343,
                                    },
                                },
                            ),
                        ],
                        loc: None,
                        range: SourceRange {
                            start: 317,
                            end: 344,
                        },
                    },
                ),
                loc: None,
                range: SourceRange {
                    start: 312,
                    end: 344,
                },
            },
        ),
        entity_type: Some(
            GenericTypeAnnotation(
                GenericTypeAnnotation {
                    id: Identifier(
                        Identifier {
                            name: "User",
                            binding: None,
                            type_annotation: None,
                            loc: None,
                            range: SourceRange {
                                start: 305,
                                end: 309,
                            },
                        },
                    ),
                    type_parameters: None,
                    loc: None,
                    range: SourceRange {
                        start: 305,
                        end: 309,
                    },
                },
            ),
        ),
        arguments: None,
        is_live: None,
    },
)

Strong(
    FieldData {
        field_name: WithLocation {
            location: <generated>:368:390,
            item: "optional_plural_string",
        },
        return_type: UnionTypeAnnotation(
            UnionTypeAnnotation {
                types: [
                    GenericTypeAnnotation(
                        GenericTypeAnnotation {
                            id: Identifier(
                                Identifier {
                                    name: "Array",
                                    binding: None,
                                    type_annotation: None,
                                    loc: None,
                                    range: SourceRange {
                                        start: 404,
                                        end: 409,
                                    },
                                },
                            ),
                            type_parameters: Some(
                                TypeParameterInstantiation {
                                    params: [
                                        StringTypeAnnotation(
                                            StringTypeAnnotation {
                                                loc: None,
                                                range: SourceRange {
                                                    start: 410,
                                                    end: 416,
                                                },
                                            },
                                        ),
                                    ],
                                    loc: None,
                                    range: SourceRange {
                                        start: 409,
                                        end: 417,
                                    },
                                },
                            ),
                            loc: None,
                            range: SourceRange {
                                start: 404,
                                end: 417,
                            },
                        },
                    ),
                    NullLiteralTypeAnnotation(
                        NullLiteralTypeAnnotation {
                            loc: None,
                            range: SourceRange {
                                start: 420,
                                end: 424,
                            },
                        },
                    ),
                    GenericTypeAnnotation(
                        GenericTypeAnnotation {
                            id: Identifier(
                                Identifier {
                                    name: "undefined",
                                    binding: None,
                                    type_annotation: None,
                                    loc: None,
                                    range: SourceRange {
                                        start: 427,
                                        end: 436,
                                    },
                                },
                            ),
                            type_parameters: None,
                            loc: None,
                            range: SourceRange {
                                start: 427,
                                end: 436,
                            },
                        },
                    ),
                ],
                loc: None,
                range: SourceRange {
                    start: 404,
                    end: 436,
                },
            },
        ),
        entity_type: Some(
            GenericTypeAnnotation(
                GenericTypeAnnotation {
                    id: Identifier(
                        Identifier {
                            name: "User",
                            binding: None,
                            type_annotation: None,
                            loc: None,
                            range: SourceRange {
                                start: 397,
                                end: 401,
                            },
                        },
                    ),
                    type_parameters: None,
                    loc: None,
                    range: SourceRange {
                        start: 397,
                        end: 401,
                    },
                },
            ),
        ),
        arguments: None,
        is_live: None,
    },
)

Strong(
    FieldData {
        field_name: WithLocation {
            location: <generated>:460:491,
            item: "optional_plural_optional_string",
        },
        return_type: UnionTypeAnnotation(
            UnionTypeAnnotation {
                types: [
                    GenericTypeAnnotation(
                        GenericTypeAnnotation {
                            id: Identifier(
                                Identifier {
                                    name: "Array",
                                    binding: None,
                                    type_annotation: None,
                                    loc: None,
                                    range: SourceRange {
                                        start: 505,
                                        end: 510,
                                    },
                                },
                            ),
                            type_parameters: Some(
                                TypeParameterInstantiation {
                                    params: [
                                        UnionTypeAnnotation(
                                            UnionTypeAnnotation {
                                                types: [
                                                    StringTypeAnnotation(
                                                        StringTypeAnnotation {
                                                            loc: None,
                                                            range: SourceRange {
                                                                start: 511,
                                                                end: 517,
                                                            },
                                                        },
                                                    ),
                                                    NullLiteralTypeAnnotation(
                                                        NullLiteralTypeAnnotation {
                                                            loc: None,
                                                            range: SourceRange {
                                                                start: 520,
                                                                end: 524,
                                                            },
                                                        },
                                                    ),
                                                    GenericTypeAnnotation(
                                                        GenericTypeAnnotation {
                                                            id: Identifier(
                                                                Identifier {
                                                                    name: "undefined",
                                                                    binding: None,
                                                                    type_annotation: None,
                                                                    loc: None,
                                                                    range: SourceRange {
                                                                        start: 527,
                                                                        end: 536,
                                                                    },
                                                                },
                                                            ),
                                                            type_parameters: None,
                                                            loc: None,
                                                            range: SourceRange {
                                                                start: 527,
                                                                end: 536,
                                                            },
                                                        },
                                                    ),
                                                ],
                                                loc: None,
                                                range: SourceRange {
                                                    start: 511,
                                                    end: 536,
                                                },
                                            },
                                        ),
                                    ],
                                    loc: None,
                                    range: SourceRange {
                                        start: 510,
                                        end: 537,
                                    },
                                },
                            ),
                            loc: None,
                            range: SourceRange {
                                start: 505,
                                end: 537,
                            },
                        },
                    ),
                    NullLiteralTypeAnnotation(
                        NullLiteralTypeAnnotation {
                            loc: None,
                            range: SourceRange {
                                start: 540,
                                end: 544,
                            },
                        },
                    ),
                    GenericTypeAnnotation(
                        GenericTypeAnnotation {
                            id: Identifier(
                                Identifier {
                                    name: "undefined",
                                    binding: None,
                                    type_annotation: None,
                                    loc: None,
                                    range: SourceRange {
                                        start: 547,
                                        end: 556,
                                    },
                                },
                            ),
                            type_parameters: None,
                            loc: None,
                            range: SourceRange {
                                start: 547,
                                end: 556,
                            },
                        },
                    ),
                ],
                loc: None,
                range: SourceRange {
                    start: 505,
                    end: 556,
                },
            },
        ),
        entity_type: Some(
            GenericTypeAnnotation(
                GenericTypeAnnotation {
                    id: Identifier(
                        Identifier {
                            name: "User",
                            binding: None,
                            type_annotation: None,
                            loc: None,
                            range: SourceRange {
                                start: 498,
                                end: 502,
                            },
                        },
                    ),
                    type_parameters: None,
                    loc: None,
                    range: SourceRange {
                        start: 498,
                        end: 502,
                    },
                },
            ),
        ),
        arguments: None,
        is_live: None,
    },
)
